src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -0,0 +1,356 @@
+"""
+usage: workspace_downloader.py [-h] --workspace_id WORKSPACE_ID [--root_dir ROOT_DIR] [--kb_base_url KB_BASE_URL] [--workers {1,2,3,4}]
+                               [--token_filename TOKEN_FILENAME] [--delete_job_dir]
+
+PROTOTYPE - Download genome files from the workspace service (WSS).
+
+options:
+  -h, --help            show this help message and exit
+
+required named arguments:
+  --workspace_id WORKSPACE_ID
+                        Workspace addressed by the permanent ID
+
+optional arguments:
+  --root_dir ROOT_DIR   Root directory.
+  --kb_base_url KB_BASE_URL
+                        KBase base URL, defaulting to prod
+  --workers             Number of workers for multiprocessing
+  --token_filename TOKEN_FILENAME
+                        Filename in home directory that stores token
+  --delete_job_dir      Delete job directory
+
+            
+e.g.
+PYTHONPATH=. python src/loaders/workspace_downloader/workspace_downloader.py --workspace_id 39795 --kb_base_url https://ci.kbase.us/services/
+
+NOTE:
+NERSC file structure for WS:
+/global/cfs/cdirs/kbase/collections/sourcedata/ -> WS -> workspace ID -> UPA -> .fa && .meta files 
+
+e.g.
+/global/cfs/cdirs/kbase/collections/sourcedata/WS -> 39795 -> 39795_10_1 -> 39795_10_1.fa 
+                                                                         -> 39795_10_1.meta
+                                                              39795_22_1 -> 39795_22_1.fa 
+                                                                         -> 39795_22_1.meta
+                                                     
+"""
+import argparse
+import itertools
+import json
+import os
+import shutil
+import stat
+import time
+from multiprocessing import Pool, Queue, cpu_count
+
+import docker
+
+from src.clients.AssemblyUtilClient import AssemblyUtil
+from src.clients.workspaceClient import Workspace
+from src.loaders.common import loader_common_names, loader_helper
+
+# setup KB_AUTH_TOKEN as env or provide a token_filename in --token_filename
+# export KB_AUTH_TOKEN="your-kb-auth-token"
+
+# supported source of data
+SOURCE = "WS"
+# filtering applied to list objects
+FILTER_OBJECTS_NAME_BY = "KBaseGenomeAnnotations.Assembly"
+# callback_port
+CALLBACK_PORT = 9999
+
+
+class Conf:
+    def __init__(self, job_dir, output_dir, workers, kb_base_url, token_filename):
+        self.setup_callback_server_envs(job_dir, kb_base_url, token_filename)
+        time.sleep(1)
+        self.start_callback_server(docker.from_env(), "cb")
```

Comment:
Could 'cb' conflict with other containers?

Reply:
I would give it a unique name just in case multiple people are running this script at the same time

Reply:
On second thought, is there any reason not to just start a new callback server container every time? Then you don't risk shutting down a container that's in use when you `stop()` it.

Reply:
Start a new callback server with an unique name every time.

Reply:
:+1: